File: services/server/src/users/controller/users.controller.ts
import {
  Controller,
  Get,
  Patch,
  Delete,
  Post,
  Param,
  Body,
  Query,
  UseGuards,
  HttpCode,
  HttpStatus,
} from "@nestjs/common";
import { ApiTags, ApiOperation, ApiBearerAuth } from "@nestjs/swagger";
import { JwtAuthGuard } from "@/auth/presentation/jwt-auth.guard";
import { CurrentUser } from "@/auth/presentation/current-user.decorator";
import { UsersService } from "@/users/application/users.service";
import { FollowService } from "@/follow/application/follow.service";
import {
  clampPageSize,
  offsetFromPageToken,
  nextOffsetPageToken,
  pageStateFromPageToken,
  nextPageStateToken,
} from "@/core/aip/page-token";

@ApiTags("users")
@Controller("users")
@UseGuards(JwtAuthGuard)
@ApiBearerAuth()
export class UsersController {
  constructor(
    private readonly usersService: UsersService,
    private readonly followService: FollowService,
  ) {}

  @Get("suggestions")
  @ApiOperation({ summary: "List suggested users to follow" })
  async getSuggestions(
    @CurrentUser() user: { id: string },
    @Query("page_size") pageSize?: string,
  ) {
    const users = await this.followService.getSuggestions(
      user.id,
      clampPageSize(pageSize ? parseInt(pageSize, 10) : undefined, 10),
    );
    return { users };
  }

  @Get()
  @ApiOperation({ summary: "List users" })
  async getUsers(
    @Query("page_size") pageSizeRaw?: string,
    @Query("page_token") pageToken?: string,
    @Query("search") search?: string,
  ) {
    const pageSize = clampPageSize(
      pageSizeRaw ? parseInt(pageSizeRaw, 10) : undefined,
    );
    const offset = offsetFromPageToken(pageToken, pageSize);
    const page = Math.floor(offset / pageSize) + 1;
    const result = await this.usersService.getUsers({
      page,
      limit: pageSize,
      ...(search && { search }),
    });
    const hasMore = offset + result.data.length < result.pagination.total;

    return {
      users: result.data,
      next_page_token: nextOffsetPageToken(offset, pageSize, hasMore),
      total_size: result.pagination.total,
    };
  }

  @Get(":user/followers")
  @ApiOperation({ summary: "List followers" })
  async getFollowers(
    @CurrentUser() user: { id: string },
    @Param("user") id: string,
    @Query("page_size") pageSizeRaw?: string,
    @Query("page_token") pageToken?: string,
  ) {
    const pageSize = clampPageSize(
      pageSizeRaw ? parseInt(pageSizeRaw, 10) : undefined,
    );
    const result = await this.followService.getFollowers(
      id,
      pageSize,
      pageStateFromPageToken(pageToken),
      user.id,
    );
    return {
      users: result.list,
      total_size: result.total,
      next_page_token: nextPageStateToken(result.pageState),
    };
  }

  @Get(":user/following")
  @ApiOperation({ summary: "List following" })
  async getFollowing(
    @CurrentUser() user: { id: string },
    @Param("user") id: string,
    @Query("page_size") pageSizeRaw?: string,
    @Query("page_token") pageToken?: string,
  ) {
    const pageSize = clampPageSize(
      pageSizeRaw ? parseInt(pageSizeRaw, 10) : undefined,
    );
    const result = await this.followService.getFollowing(
      id,
      pageSize,
      pageStateFromPageToken(pageToken),
      user.id,
    );
    return {
      users: result.list,
      total_size: result.total,
      next_page_token: nextPageStateToken(result.pageState),
    };
  }

  @Get(":user")
  @ApiOperation({ summary: "Get user" })
  async getUser(@Param("user") id: string) {
    const profile = await this.usersService.getUserById(id);
    const [followersCount, followingCount] = await Promise.all([
      this.followService.getFollowersCount(id),
      this.followService.getFollowingCount(id),
    ]);

    return { ...profile, followersCount, followingCount };
  }

  @Patch(":user")
  @ApiOperation({ summary: "Update user" })
  async updateUser(
    @Param("user") id: string,
    @Body()
    updateData: {
      username?: string;
      email?: string;
      phone?: string;
      avatar?: string;
      status?: string;
    },
  ) {
    return this.usersService.updateUser(id, updateData);
  }

  @Delete(":user")
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: "Delete user" })
  async deleteUser(@Param("user") id: string) {
    await this.usersService.deleteUser(id);
  }

  @Post("following:check")
  @ApiOperation({ summary: "Batch check following state" })
  async checkFollowing(
    @CurrentUser() user: { id: string },
    @Body() body: { userIds?: string[] },
  ) {
    const ids = Array.isArray(body?.userIds) ? body.userIds : [];
    const set = await this.followService.checkFollowing(user.id, ids);
    return {
      results: ids.map((id) => ({ userId: id, isFollowing: set.has(id) })),
    };
  }

  @Post(":user\\:block")
  @ApiOperation({ summary: "Block user" })
  async blockUser(
    @CurrentUser() user: { id: string },
    @Param("user") blockedId: string,
  ) {
    await this.usersService.blockUser(user.id, blockedId);
    return {};
  }

  @Post(":user\\:unblock")
  @ApiOperation({ summary: "Unblock user" })
  async unblockUser(
    @CurrentUser() user: { id: string },
    @Param("user") blockedId: string,
  ) {
    await this.usersService.unblockUser(user.id, blockedId);
    return {};
  }
}

File: src/main/java/com/chat/users/controller/UsersController.java
package com.chat.users.controller;

import com.chat.users.service.UsersService;
import java.util.List;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/users")
public class UsersController {

  private final UsersService usersService;

  public UsersController(UsersService usersService) {
    this.usersService = usersService;
  }

  @GetMapping("me")
  public Map<String, UserResponse> me(Authentication authentication) {
    return Map.of("user", usersService.getById(authentication.getName()));
  }

  @GetMapping("suggestions")
  public Map<String, Object> suggestions(
      Authentication authentication,
      @RequestParam(value = "page_size", required = false) Integer pageSize) {
    int size = pageSize == null ? 10 : Math.min(pageSize, 100);
    return usersService.suggestions(authentication.getName(), size);
  }

  @PostMapping("following:check")
  public Map<String, Object> followingCheck(
      Authentication authentication, @RequestBody Map<String, List<String>> body) {
    List<String> userIds = body.getOrDefault("userIds", List.of());
    return usersService.checkFollowing(authentication.getName(), userIds);
  }

  @GetMapping
  public Map<String, List<UserResponse>> search(
      @RequestParam(value = "q", required = false) String query,
      @RequestParam(value = "page_size", required = false) Integer pageSize) {
    int size = pageSize == null ? 20 : Math.min(pageSize, 100);
    List<UserResponse> users =
        query == null || query.isBlank() ? List.of() : usersService.search(query, size);
    return Map.of("users", users);
  }

  @GetMapping("{user}")
  public UserResponse get(@PathVariable("user") String userId) {
    return usersService.getById(userId);
  }

  @PostMapping("{user}:block")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public void block(Authentication authentication, @PathVariable("user") String userId) {
    usersService.block(authentication.getName(), userId);
  }

  @PostMapping("{user}:unblock")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public void unblock(Authentication authentication, @PathVariable("user") String userId) {
    usersService.unblock(authentication.getName(), userId);
  }

  @DeleteMapping("me")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public void deleteMe(Authentication authentication) {
    usersService.delete(authentication.getName());
  }
}
